File: server/services/enrollment_service.ts
import { db } from '../db';

export function getStudentClasses(studentId: string) {
  const query = `
    SELECT 
      c.class_id,
      s.subject_id,
      s.short_name AS subject_short,
      s.full_name AS subject_full,
      COALESCE(s.year_level, 1) AS year_level,
      p.full_name AS professor,
      g.group_name,
      c.room,
      e.status,
      e.enrolled_at,
      e.dropped_at
    FROM student_class_enrollment e
    JOIN classes c ON e.class_id = c.class_id
    JOIN subjects s ON c.subject_id = s.subject_id
    JOIN professors p ON c.professor_id = p.professor_id
    JOIN groups g ON c.group_id = g.group_id
    WHERE e.student_id = ?
    ORDER BY 
      CASE WHEN e.status = 'active' THEN 0 ELSE 1 END,
      e.enrolled_at DESC
  `;
  const allRows = db.query(query, [studentId]).rows;
  
  // Deduplicate: Each unique subject_id MUST appear only once in the Courses view!
  const subjectMap = new Map<number, any>();
  for (const row of allRows) {
    if (!subjectMap.has(row.subject_id)) {
      subjectMap.set(row.subject_id, row);
    }
  }
  return Array.from(subjectMap.values()).sort((a, b) => a.subject_full.localeCompare(b.subject_full));
}

export function dropClass(studentId: string, classId: number) {
  // 1. Find subject_id
  const cls = db.query('SELECT subject_id FROM classes WHERE class_id = ?', [classId]).rows[0];
  const now = new Date().toISOString();

  if (cls) {
    // Drop all enrollment rows for this subject
    db.query(
      `UPDATE student_class_enrollment 
       SET status = 'dropped', dropped_at = ? 
       WHERE student_id = ? AND class_id IN (SELECT class_id FROM classes WHERE subject_id = ?)`,
      [now, studentId, cls.subject_id]
    );

    // Remove any schedule override for this subject so it is removed from the timetable
    db.query(
      `DELETE FROM student_schedule_overrides 
       WHERE student_id = ? AND class_id IN (SELECT class_id FROM classes WHERE subject_id = ?)`,
      [studentId, cls.subject_id]
    );
  } else {
    db.query(
      `UPDATE student_class_enrollment 
       SET status = 'dropped', dropped_at = ? 
       WHERE student_id = ? AND class_id = ?`,
      [now, studentId, classId]
    );
  }

  // 3. Cascade to homework submissions: deactivate submissions where deadline > now
  const homeworks = db.query(
    `SELECT h.homework_id 
     FROM homeworks h
     JOIN homework_submissions hs ON h.homework_id = hs.homework_id
     WHERE h.class_id = ? AND hs.student_id = ? AND h.deadline > ?`,
    [classId, studentId, now]
  ).rows;

  if (homeworks.length > 0) {
    const hwIds = homeworks.map((h) => h.homework_id);
    const placeholders = hwIds.map(() => '?').join(',');
    db.query(
      `UPDATE homework_submissions 
       SET is_active = 0 
       WHERE student_id = ? AND homework_id IN (${placeholders})`,
      [studentId, ...hwIds]
    );
  }

  return {
    success: true,
    message: 'Course successfully dropped',
    removed_homeworks_count: homeworks.length,
  };
}

export function retakeClass(studentId: string, classId: number) {
  // 1. Update enrollment status to active
  db.query(
    `UPDATE student_class_enrollment 
     SET status = 'active', dropped_at = NULL 
     WHERE student_id = ? AND class_id = ?`,
    [studentId, classId]
  );

  const student = db.query('SELECT group_id FROM students WHERE student_id = ?', [studentId]).rows[0];
  const cls = db.query('SELECT class_id, group_id, subject_id FROM classes WHERE class_id = ?', [classId]).rows[0];

  // If class is from a different group, ensure schedule slot is present in student_schedule_overrides
  if (student && cls && student.group_id !== cls.group_id) {
    const slots = db.query(
      `SELECT day_of_week, start_time, end_time FROM group_timetable WHERE class_id = ?`,
      [classId]
    ).rows;
    for (const slot of slots) {
      db.query(
        `INSERT INTO student_schedule_overrides (student_id, day_of_week, start_time, end_time, class_id)
         VALUES (?, ?, ?, ?, ?)`,
        [studentId, slot.day_of_week, slot.start_time, slot.end_time, classId]
      );
    }
  }

  // 2. Cascade to homework submissions: restore submissions where deadline > now
  const now = new Date().toISOString();
  const openHomeworks = db.query(
    `SELECT homework_id 
     FROM homeworks 
     WHERE class_id = ? AND deadline > ?`,
    [classId, now]
  ).rows;

  let restoredCount = 0;
  for (const hw of openHomeworks) {
    const existing = db.query(
      `SELECT submission_id FROM homework_submissions WHERE homework_id = ? AND student_id = ?`,
      [hw.homework_id, studentId]
    ).rows;

    if (existing.length > 0) {
      db.query(
        `UPDATE homework_submissions SET is_active = 1 WHERE submission_id = ?`,
        [existing[0].submission_id]
      );
      restoredCount++;
    } else {
      db.query(
        `INSERT INTO homework_submissions (homework_id, student_id, is_done, is_active) VALUES (?, ?, 0, 1)`,
        [hw.homework_id, studentId]
      );
      restoredCount++;
    }
  }

  return {
    success: true,
    message: 'Course successfully re-enrolled',
    restored_homeworks_count: restoredCount,
  };
}

export function getRetakeCatalog(studentId: string) {
  const student = db.query(
    `SELECT student_id, full_name, group_id, COALESCE(year_of_study, 2) AS year_of_study FROM students WHERE student_id = ?`,
    [studentId]
  ).rows[0];

  if (!student) {
    throw new Error(`Student ${studentId} not found`);
  }

  const userYear = Number(student.year_of_study) || 2;
  // Available course years: all academic year levels <= user's year
  // e.g. Year 2 student can take Year 1 (lower) or Year 2 (current)
  const availableYears: number[] = [];
  for (let y = 1; y <= userYear; y++) {
    availableYears.push(y);
  }

  const placeholders = availableYears.map(() => '?').join(',');
  const subjects = db.query(
    `SELECT 
       s.subject_id,
       s.short_name,
       s.full_name,
       COALESCE(s.year_level, 1) AS year_level
     FROM subjects s
     WHERE COALESCE(s.year_level, 1) IN (${placeholders})
     ORDER BY s.year_level ASC, s.full_name ASC`,
    availableYears
  ).rows;

  // Check enrollment status for each subject for this student
  const activeEnrollments = db.query(
    `SELECT c.subject_id, e.status, e.class_id 
     FROM student_class_enrollment e
     JOIN classes c ON e.class_id = c.class_id
     WHERE e.student_id = ?`,
    [studentId]
  ).rows;

  const subjectEnrollmentMap: Record<number, { status: string; class_id: number }> = {};
  for (const row of activeEnrollments) {
    // If student has multiple enrollments, prefer 'active'
    if (!subjectEnrollmentMap[row.subject_id] || row.status === 'active') {
      subjectEnrollmentMap[row.subject_id] = {
        status: row.status,
        class_id: row.class_id,
      };
    }
  }

  const enrichedSubjects = subjects.map((sub) => {
    const enr = subjectEnrollmentMap[sub.subject_id];
    return {
      subject_id: sub.subject_id,
      short_name: sub.short_name,
      full_name: sub.full_name,
      year_level: sub.year_level,
      enrollment_status: enr ? enr.status : 'none',
      is_enrolled: enr ? enr.status === 'active' : false,
      current_class_id: enr ? enr.class_id : null,
    };
  });

  return {
    student_id: student.student_id,
    student_name: student.full_name,
    student_year: userYear,
    available_years: availableYears,
    subjects: enrichedSubjects,
  };
}

export function enrollRetakeClass(studentId: string, classId: number) {
  const cls = db.query(
    `SELECT c.class_id, c.subject_id, c.group_id, s.full_name AS subject_name, s.short_name
     FROM classes c
     JOIN subjects s ON c.subject_id = s.subject_id
     WHERE c.class_id = ?`,
    [classId]
  ).rows[0];

  if (!cls) {
    throw new Error('Selected class section was not found');
  }

  const student = db.query('SELECT student_id, group_id FROM students WHERE student_id = ?', [studentId]).rows[0];
  if (!student) {
    throw new Error('Student not found');
  }

  // 1. Update or Insert enrollment
  const existingEnrollment = db.query(
    `SELECT enrollment_id FROM student_class_enrollment WHERE student_id = ? AND class_id = ?`,
    [studentId, classId]
  ).rows;

  if (existingEnrollment.length > 0) {
    db.query(
      `UPDATE student_class_enrollment SET status = 'active', dropped_at = NULL WHERE student_id = ? AND class_id = ?`,
      [studentId, classId]
    );
  } else {
    // If student had an enrollment in ANOTHER class for the same subject, deactivate it
    const sameSubjectEnrollments = db.query(
      `SELECT e.class_id FROM student_class_enrollment e
       JOIN classes c ON e.class_id = c.class_id
       WHERE e.student_id = ? AND c.subject_id = ?`,
      [studentId, cls.subject_id]
    ).rows;
    for (const se of sameSubjectEnrollments) {
      db.query(`UPDATE student_class_enrollment SET status = 'dropped' WHERE student_id = ? AND class_id = ?`, [studentId, se.class_id]);
    }

    db.query(
      `INSERT INTO student_class_enrollment (student_id, class_id, status) VALUES (?, ?, 'active')`,
      [studentId, classId]
    );
  }

  // 2. Clear old schedule overrides for this subject
  db.query(
    `DELETE FROM student_schedule_overrides 
     WHERE student_id = ? AND class_id IN (SELECT class_id FROM classes WHERE subject_id = ?)`,
    [studentId, cls.subject_id]
  );

  // 3. If class is not in student's base group, add to student_schedule_overrides
  const slots = db.query(
    `SELECT day_of_week, start_time, end_time FROM group_timetable WHERE class_id = ?`,
    [classId]
  ).rows;

  for (const slot of slots) {
    db.query(
      `INSERT INTO student_schedule_overrides (student_id, day_of_week, start_time, end_time, class_id)
       VALUES (?, ?, ?, ?, ?)`,
      [studentId, slot.day_of_week, slot.start_time, slot.end_time, classId]
    );
  }

  // 4. Cascade active homeworks for this class
  const now = new Date().toISOString();
  const openHomeworks = db.query(
    `SELECT homework_id FROM homeworks WHERE class_id = ? AND deadline > ?`,
    [classId, now]
  ).rows;

  let assignedCount = 0;
  for (const hw of openHomeworks) {
    const exists = db.query(
      `SELECT submission_id FROM homework_submissions WHERE homework_id = ? AND student_id = ?`,
      [hw.homework_id, studentId]
    ).rows;
    if (exists.length > 0) {
      db.query(`UPDATE homework_submissions SET is_active = 1 WHERE submission_id = ?`, [exists[0].submission_id]);
    } else {
      db.query(`INSERT INTO homework_submissions (homework_id, student_id, is_done, is_active) VALUES (?, ?, 0, 1)`, [hw.homework_id, studentId]);
    }
    assignedCount++;
  }

  return {
    success: true,
    message: `Successfully registered for retake course: ${cls.subject_name} (${cls.short_name})`,
    assigned_homeworks_count: assignedCount,
  };
}

File: backend/services/enrollment_service.py
from typing import List, Dict, Any
from datetime import datetime
from database import query, execute

def get_student_classes(student_id: str) -> List[Dict[str, Any]]:
    all_rows = query(
        """SELECT 
               c.class_id,
               s.subject_id,
               s.short_name AS subject_short,
               s.full_name AS subject_full,
               COALESCE(s.year_level, 1) AS year_level,
               p.full_name AS professor,
               g.group_name,
               c.room,
               e.status,
               e.enrolled_at,
               e.dropped_at
           FROM student_class_enrollment e
           JOIN classes c ON e.class_id = c.class_id
           JOIN subjects s ON c.subject_id = s.subject_id
           JOIN professors p ON c.professor_id = p.professor_id
           JOIN groups g ON c.group_id = g.group_id
           WHERE e.student_id = ?
           ORDER BY 
               CASE WHEN e.status = 'active' THEN 0 ELSE 1 END,
               e.enrolled_at DESC""",
        (student_id,)
    )

    subject_map: Dict[int, Dict[str, Any]] = {}
    for row in all_rows:
        sid = row["subject_id"]
        if sid not in subject_map:
            subject_map[sid] = row

    result = list(subject_map.values())
    result.sort(key=lambda x: x["subject_full"])
    return result

def drop_class(student_id: str, class_id: int) -> Dict[str, Any]:
    cls_rows = query("SELECT subject_id FROM classes WHERE class_id = ?", (class_id,))
    now_iso = datetime.now().isoformat()

    if cls_rows:
        subject_id = cls_rows[0]["subject_id"]
        execute(
            """UPDATE student_class_enrollment 
               SET status = 'dropped', dropped_at = ? 
               WHERE student_id = ? AND class_id IN (SELECT class_id FROM classes WHERE subject_id = ?)""",
            (now_iso, student_id, subject_id)
        )
        execute(
            """DELETE FROM student_schedule_overrides 
               WHERE student_id = ? AND class_id IN (SELECT class_id FROM classes WHERE subject_id = ?)""",
            (student_id, subject_id)
        )
    else:
        execute(
            """UPDATE student_class_enrollment 
               SET status = 'dropped', dropped_at = ? 
               WHERE student_id = ? AND class_id = ?""",
            (now_iso, student_id, class_id)
        )

    # Homeworks cascade
    homeworks = query(
        """SELECT h.homework_id 
           FROM homeworks h
           JOIN homework_submissions hs ON h.homework_id = hs.homework_id
           WHERE h.class_id = ? AND hs.student_id = ? AND h.deadline > ?""",
        (class_id, student_id, now_iso)
    )
    if homeworks:
        hw_ids = [h["homework_id"] for h in homeworks]
        placeholders = ",".join("?" for _ in hw_ids)
        execute(
            f"UPDATE homework_submissions SET is_active = 0 WHERE student_id = ? AND homework_id IN ({placeholders})",
            tuple([student_id] + hw_ids)
        )

    return {
        "success": True,
        "message": "Course successfully dropped",
        "removed_homeworks_count": len(homeworks),
    }

def retake_class(student_id: str, class_id: int) -> Dict[str, Any]:
    execute(
        """UPDATE student_class_enrollment 
           SET status = 'active', dropped_at = NULL 
           WHERE student_id = ? AND class_id = ?""",
        (student_id, class_id)
    )

    student_rows = query("SELECT group_id FROM students WHERE student_id = ?", (student_id,))
    cls_rows = query("SELECT class_id, group_id, subject_id FROM classes WHERE class_id = ?", (class_id,))

    if student_rows and cls_rows and student_rows[0]["group_id"] != cls_rows[0]["group_id"]:
        slots = query("SELECT day_of_week, start_time, end_time FROM group_timetable WHERE class_id = ?", (class_id,))
        for slot in slots:
            execute(
                """INSERT INTO student_schedule_overrides (student_id, day_of_week, start_time, end_time, class_id)
                   VALUES (?, ?, ?, ?, ?)""",
                (student_id, slot["day_of_week"], slot["start_time"], slot["end_time"], class_id)
            )

    now_iso = datetime.now().isoformat()
    open_homeworks = query(
        "SELECT homework_id FROM homeworks WHERE class_id = ? AND deadline > ?",
        (class_id, now_iso)
    )

    restored = 0
    for hw in open_homeworks:
        existing = query(
            "SELECT submission_id FROM homework_submissions WHERE homework_id = ? AND student_id = ?",
            (hw["homework_id"], student_id)
        )
        if existing:
            execute("UPDATE homework_submissions SET is_active = 1 WHERE submission_id = ?", (existing[0]["submission_id"],))
        else:
            execute(
                "INSERT INTO homework_submissions (homework_id, student_id, is_done, is_active) VALUES (?, ?, 0, 1)",
                (hw["homework_id"], student_id)
            )
        restored += 1

    return {
        "success": True,
        "message": "Course successfully re-enrolled",
        "restored_homeworks_count": restored,
    }

def get_retake_catalog(student_id: str) -> Dict[str, Any]:
    student_rows = query(
        "SELECT student_id, full_name, group_id, COALESCE(year_of_study, 2) AS year_of_study FROM students WHERE student_id = ?",
        (student_id,)
    )
    if not student_rows:
        raise ValueError(f"Student {student_id} not found")
    student = student_rows[0]

    user_year = int(student.get("year_of_study") or 2)
    available_years = list(range(1, user_year + 1))
    placeholders = ",".join("?" for _ in available_years)

    subjects = query(
        f"""SELECT 
               s.subject_id,
               s.short_name,
               s.full_name,
               COALESCE(s.year_level, 1) AS year_level
           FROM subjects s
           WHERE COALESCE(s.year_level, 1) IN ({placeholders})
           ORDER BY s.year_level ASC, s.full_name ASC""",
        tuple(available_years)
    )

    active_enrollments = query(
        """SELECT c.subject_id, e.status, e.class_id 
           FROM student_class_enrollment e
           JOIN classes c ON e.class_id = c.class_id
           WHERE e.student_id = ?""",
        (student_id,)
    )

    subject_enr_map = {}
    for row in active_enrollments:
        sid = row["subject_id"]
        if sid not in subject_enr_map or row["status"] == "active":
            subject_enr_map[sid] = row

    enriched = []
    for sub in subjects:
        sid = sub["subject_id"]
        enr = subject_enr_map.get(sid)
        is_enr = enr["status"] == "active" if enr else False
        enriched.append({
            "subject_id": sid,
            "short_name": sub["short_name"],
            "full_name": sub["full_name"],
            "year_level": sub["year_level"],
            "enrollment_status": enr["status"] if enr else "none",
            "is_enrolled": is_enr,
            "current_class_id": enr["class_id"] if enr else None,
        })

    return {
        "student_id": student["student_id"],
        "student_name": student["full_name"],
        "student_year": user_year,
        "available_years": available_years,
        "subjects": enriched,
    }

def enroll_retake_class(student_id: str, class_id: int) -> Dict[str, Any]:
    cls_rows = query(
        """SELECT c.class_id, c.subject_id, c.group_id, s.full_name AS subject_name, s.short_name
           FROM classes c
           JOIN subjects s ON c.subject_id = s.subject_id
           WHERE c.class_id = ?""",
        (class_id,)
    )
    if not cls_rows:
        raise ValueError(f"Class {class_id} not found")
    cls = cls_rows[0]

    existing = query(
        "SELECT enrollment_id, status FROM student_class_enrollment WHERE student_id = ? AND class_id = ?",
        (student_id, class_id)
    )
    if existing:
        execute(
            "UPDATE student_class_enrollment SET status = 'active', dropped_at = NULL WHERE student_id = ? AND class_id = ?",
            (student_id, class_id)
        )
    else:
        execute(
            "INSERT INTO student_class_enrollment (student_id, class_id, status) VALUES (?, ?, 'active')",
            (student_id, class_id)
        )

    # If section belongs to another group, add schedule slots
    student_rows = query("SELECT group_id FROM students WHERE student_id = ?", (student_id,))
    if student_rows and student_rows[0]["group_id"] != cls["group_id"]:
        slots = query("SELECT day_of_week, start_time, end_time FROM group_timetable WHERE class_id = ?", (class_id,))
        for slot in slots:
            execute(
                """INSERT INTO student_schedule_overrides (student_id, day_of_week, start_time, end_time, class_id)
                   VALUES (?, ?, ?, ?, ?)""",
                (student_id, slot["day_of_week"], slot["start_time"], slot["end_time"], class_id)
            )

    return {
        "success": True,
        "message": f"Successfully enrolled into {cls['subject_name']}!",
        "subject_name": cls["subject_name"],
        "class_id": class_id,
    }
